File: research/python-lab/src/edgechain_lab/mars/spatial.py
"""Spatial jury scoring."""

from __future__ import annotations

import numpy as np

from edgechain_lab.fl.metrics import cosine_similarity


def spatial_jury_score(
    site_id: str,
    updates_by_site: dict[str, np.ndarray],
    cluster_by_site: dict[str, str],
) -> float:
    """Compare one site update to its cluster peer consensus."""

    cluster_id = cluster_by_site[site_id]
    peers = [
        update
        for peer_site, update in updates_by_site.items()
        if peer_site != site_id and cluster_by_site.get(peer_site) == cluster_id
    ]
    if not peers:
        return 0.5

    consensus = np.mean(peers, axis=0)
    score = max(0.0, cosine_similarity(updates_by_site[site_id], consensus))
    return float(np.clip(score, 0.0, 1.0))

File: packages/mars/src/spatial.ts
import { clamp, cosineSimilarity, vectorMean } from './math';

export function spatialJuryScore(
  siteId: string,
  updatesBySite: Record<string, number[]>,
  clusterBySite: Record<string, string>,
): number {
  const clusterId = clusterBySite[siteId];
  const peers = Object.entries(updatesBySite)
    .filter(([peerSite]) => peerSite !== siteId && clusterBySite[peerSite] === clusterId)
    .map(([, update]) => update);

  if (peers.length === 0) {
    return 0.5;
  }

  const consensus = vectorMean(peers);
  return clamp(Math.max(0, cosineSimilarity(updatesBySite[siteId], consensus)));
}
